### ISSUE TITLE
Wrong endpoint used in metadata when specifying viewset

### ISSUE THREAD

**Issue by xnegativx** [drf-forms/drf-schema-adapter#35]
Hi,
When an endpoint is registered with a specific viewset, a different endpoint is generated in determine_metadata (hence losing all out custom attributes). The responsible part in metada.py:
```
        try:
            serializer_instance = view.get_serializer()
        except Exception:
            # Custom viewset is expecting something we can't guess
            serializer_instance = view.get_serializer_class()()
        endpoint = None
        if hasattr(view, 'endpoint'):
            endpoint = view.endpoint
        else:
            if hasattr(view.get_serializer_class().Meta, 'model'):
                from .endpoints import Endpoint
                endpoint = Endpoint(view.get_serializer_class().Meta.model, viewset=view)
```
The endpoint relation is normally set by the viewset_factory, and it's not when specifying a viewset. 
Came up with this, not ideal, fix :
```
class Endpoint(with_metaclass(EndpointMetaClass, BaseEndpoint)):

    def __init__(self, model=None, **kwargs):
        self.inflector = Inflector(self.inflector_language)

        if model is not None:
            self.model = model

        arg_names = ('fields', 'serializer', 'permission_classes', 'filter_fields', 'search_fields',
                     'viewset', 'read_only', 'include_str', 'ordering_fields', 'page_size',
                     'base_viewset', 'fields_annotation', 'fieldsets', 'base_serializer', 'list_me')
        for arg_name in arg_names:
            setattr(self, arg_name, kwargs.pop(arg_name, getattr(self, arg_name, None)))
            if arg_name == 'viewset' and getattr(self, arg_name):
                assert getattr(self.viewset, 'endpoint', None) is None, 'This viewset is already registered by an endpoint'
                setattr(self.viewset, 'endpoint', self)
```
I think we have to prevent the same viewset to be registered by different endpoints or the metadata woulnd't mean anything. We could also test the class instead of the objects.. (The tests will not pass as the same viewset is registered multiple times for now)

**Comment by nanuxbe** [comment]
I think there might e several ways to fix this.

First of all, it might simply be a matter of making the documentation clearer. When I want to use a custom ViewSet, 99.9% of the time, I create a "partial" ViewSet with the extra things I need and use that as a `base_viewset`on the Endpoint. I don't know if you tried that. If you didn't know it was a (recommended) option, it means the documentation should make it more obvious. I'm planning on merging https://github.com/drf-forms/drf-schema-adapter/blob/cookbooks/docs/cookbooks/endpoints.md soon, maybe it is enough?
Can you tell me more about your use case?

Second, your solution seems reasonable.

Third, the metadata is creating the Endpoint because it assumes there is None. I don't remember off-hand if the router registers itself on the ViewSet in any way but there might be a simple way to check that "something" with the same url exists and if that something is an Enpoint then that Endpoint should be used instead of creating a new one

**Comment by xnegativx** [comment]
Unfortunately using base_viewset is not completly solving my issue.
My viewset has the method get_serializer_class overriden to use different serializers for read/write operations. I get the serializer I want using the action attribute of the view (create/update/list or metadata in our case, default to list serializer).
When using the base_viewset, the wrong serializer will be used when determining metadata as it was instantiated along with the endpoint (so without the action).
I have the same issue actually even when using viewset which I solved by doing this in determine_metadata:
```
        if hasattr(view, 'endpoint'):
            endpoint = view.endpoint
            endpoint.serializer = view.get_serializer_class()
```
The same trick is not fully working with base_viewset, as the base list of fields is wrong. (There's no serializer defined when getting the list so all the fields are fetched), so in get_field_dict even if the serializer is correct there I end up will all the fields anyway because of this part:
```
    try:
        field_instance = serializer_instance.fields[name]
    except KeyError:
        return {'key': name}
```
Though I can still filter them in the adapter then because the type is not set then.

Summing up:
- With a dynamic get_queryset_class on a viewset, the wrong serializer is used when determining_metadata (using viewset or base_viewset)
- When using base_viewset the list of fields is also wrongly generated (but can be handled in the adapter)
- When using a viewset a fresh new endpoint is generated to determine_metadata, losing the customisation of the original one (original issue)

**Comment by nanuxbe** [comment]
Ok,

So I do that kind of things pretty often but it's usually to use a different queryset/serializer for detail views and list views, or use a "limited" serializer depending on the user's rights. I think it's probably not completely the same use-case though.

I'd like to understand the major differences between your serializers, my go-to solution for tis kind of problems would be to use read-only and write-only fields on the serializer but I guess if you're not using that solution is because you have a good reason not to.<
So does one of the serializer have all the fields and another one not (this is my usecase with list/detail)? And can therefore the serializer with all fields be used "by default"?

I guess my main confusion is about the way you use metadata. Te way it was intended is for use on the "OPTIONS" call which doesn't know about whether you are going to be using the info for reading or for writing. But it seems you're using it for something else?

So, summing it up:
1 & 2 - I think I need more info
3 - Your original solution is probably the best way to go for now

**Comment by xnegativx** [comment]
Well, I'm using the metadata only for generating the forms on the frontend, so for writing only. 

> So does one of the serializers have all the fields and another one not (this is my use case with list/detail)? And can therefore the serializer with all fields be used "by default"?
 Yes it's exactly what's hapenning. 
I guess that rewriting my serializers with write only/read only might solve most of the issues. But I have a case when the field is writable only at creation time. (Which my solution is not covering anyway). I'll dig on this case, I might better have my own metadatamixin closer to the original DRF one with fields definition by action

**Comment by nanuxbe** [comment]
So something along the line of

```pyhton
class MyViewSet(viewsets.ModelViewSet):

    def get_serializer_class(self):
        if getattr(self, action, None) not in ('update', 'create'):
            return MyReadSerializer
        return super(MyViewSet, self).get_serializer_class()
        
    
class MyEndpoint(Endpoint):
    (base_)serializer = MyWriteSerializer
    base_viewset = MyViewSet
```

For the case for the case of the field that's only writeable at creation time, I'm afraid you'll have to write some frontend code tough

**Comment by nanuxbe** [comment]
An example of how you can work out the "create only" issue would be to add a `create_only_fields` attribute to your Endpoint and use that from the adapter. 
ie:

```python
class MyViewSet(viewsets.ModelViewSet):

    def get_serializer_class(self):
        if getattr(self, action, None) in ('list', 'retrieve'):
            return MyReadSerializer
        if getattr(self, action, None) == 'update':
            return MyUpdateSerializer
        return super(MyViewSet, self).get_serializer_class()
        
    
class MyEndpoint(Endpoint):
    (base_)serializer = MyCreateSerializer
    base_viewset = MyViewSet


class MyAdapter(BaseAdapter):
    metadata_info = [
        ...
        MetaDataInfo('create_only_fields', PROPERTY, []),
    ]
```

The in your for generation code, if the record is not a new one, you can check that the field is not listed in  `create_only_fields` before rendering it.

**Comment by nanuxbe** [comment]
As there is no more activity on this issue, I'm assuming the proposed method works for you and I'm closing the issue.
If this is not the case, please feel free to re-open
